test_build_tag_push.py

```diff
@@ -0,0 +1,99 @@
+import unittest
+import build_tag_push
+import os
+import subprocess
+import argparse
+from unittest.mock import patch
+import io
+import sys
+
+path_to_test_yml = ('/Users/alyssakelley/Desktop/MVPStudio/build_tag_push/'
+                    'test_data/builds_succeeds_example/app.yml')
+
+path_to_fail_test_yml = ('/Users/alyssakelley/Desktop/MVPStudio/'
+                         'build_tag_push/test_data/build_fails_example/'
+                         'app.yml')
+
+
+class TestBuildTagPush(unittest.TestCase):
+
+    def test_parse_app_file(self):
+
+        data_from_app_yml = build_tag_push.parse_app_file(path_to_test_yml)
+        self.assertEqual(data_from_app_yml['name'], 'build-setup-test')
+        self.assertEqual(data_from_app_yml['build'], 'build.sh')
+        self.assertEqual(data_from_app_yml['test'], 'test.sh')
+        self.assertEqual(data_from_app_yml['docker_context'],
+                         ['build', 'static', 'Dockerfile'])
+
+    def test_build_check_app_throws_on_build(self):
+        data_from_app_yml = build_tag_push.parse_app_file(path_to_test_yml)
+
+        with self.assertRaises(
+             build_tag_push.BuildFailedException) as shouldThrow:
+
+            build_tag_push.build_check_app(data_from_app_yml,
+                                           path_to_fail_test_yml, True)
+
+        self.assertEqual("The build failed.", shouldThrow.exception.message)
+
+    def test_build_check_app_runs_build(self):
+        data_from_app_yml = build_tag_push.parse_app_file(path_to_test_yml)
+        build_tag_push.build_check_app(data_from_app_yml, path_to_test_yml,
+                                       True)
+        self.assertTrue(os.path.exists('/Users/alyssakelley/Desktop/MVPStudio/'
+                        'build_tag_push/test_data/builds_succeeds_example/tmp/'
+                                       'build_done.sh'))
+
+    def test_create_context_directory(self):
+
+        context_array = ['build', 'static', 'Dockerfile']
+        test_path = build_tag_push.create_context_directory(context_array,
+                                                            path_to_test_yml)
+        output = os.listdir(test_path)
+        self.assertEqual(output, ['Dockerfile', 'static', 'build'])
+
+    def test_build_docker_image(self):
+        data_from_app_yml = build_tag_push.parse_app_file(path_to_test_yml)
+
+        capturedOutput = io.StringIO()
+        sys.stdout = capturedOutput
```

Comment:
This is a cool trick (impressed you thought of it) but it's a bit brittle and hacky -- brittle because if line 61 throws stdout is broken for the rest of your tests.

What we really want to test here is that the docker image was built. So why not do this:

1. run `docker rmi` to remove the docker image your test creates at the beginning of the test.
2. Call `build_docker_image`
3. Do a `subprocess.call(['docker', `image`, `inspect`, image_hash])` to get info about the image your just built. if the image doesn't exist `docker image inspect` will return a non-zero exit code.

Reply:
Thanks! Good to know about stdout potentially breaking though. For this test, should I be including the image_hash even though that is being tested in the next test function? For example, my lines for this function are:

        does_image_exist = subprocess.call(['docker', 'inspect', image_name])

        self.assertEqual(does_image_exist, 0)

and the lines for the test_tag_docker_imager are:

        test_current_hash = subprocess.check_output(
                            ["git", "log", "-1", "--pretty=%h"]).strip(
                                ).decode('ascii')
        is_imaged_tagged = subprocess.call(['docker', 'image',
                                           'inspect', docker_hub_name + '/' +
                                            image_name + ':' +
                                            test_current_hash])

        self.assertEqual(is_imaged_tagged, 0)

also, as you can see in the code I included, I set the return code to a variable and then asserted that it was 0, is this ok to have?